sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1033,9 +1033,12 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
         requiredAttrIds.contains(a.exprId)) =>
         s.withMetadataColumns()
       case p: Project if p.metadataOutput.exists(a => requiredAttrIds.contains(a.exprId)) =>
+        // Inject the requested metadata columns into the project's output, if not already present.
```

Comment:
do we hit a real issue with this? If the metadata col is already in the output, this code path should not be triggered as we have `val metaCols = getMetadataAttributes(node).filterNot(inputAttrs.contains)`

What we can improve is to only include metadata columns that are included in `requiredAttrIds`.

Reply:
I hit a weird endless loop with this while debugging this `SubqueryAlias` issue. Basically, if the plan root already has a metadata attribute, but it's not available because the `SubqueryAlias` blocked it, this rule kept endlessly (re)appending the metadata column to the projections below the `SubqueryAlias`. Once the rule ran 100 times (leaving 100 copies of `_metadata` in the `Project` output), the endless loop detector kicked in and killed it.

Reply:
Re the "only include" comment, do you mean something like this?
```scala
val missingMetadata = p.metadataOutput
  .filter(a => requiredAttrIds.contains(a.exprId)
  .filterNot(a => p.projectList.exists(_.exprId == a.exprId))
```

Reply:
yup

Reply:
> but it's not available because the SubqueryAlias blocked it, this rule kept endlessly (re)appending the metadata column

Make sense. So this change is a safe guard